core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java

```diff
@@ -79,6 +82,37 @@ public Double getDistinctRowCount(RelNode rel, RelMetadataQuery mq,
     return null;
   }
 
+  public Double getDistinctRowCount(TableScan rel, RelMetadataQuery mq,
+      ImmutableBitSet groupKey, RexNode predicate) {
+    final double selectivity = predicate == null ? 1D : mq.getSelectivity(rel, predicate);
+    final boolean uniq = RelMdUtil.areColumnsDefinitelyUnique(mq, rel, groupKey);
+    if (uniq) {
+      return NumberUtil.multiply(mq.getRowCount(rel), selectivity);
+    } else if (groupKey.isEmpty()) {
+      return null;
+    } else {
+      final List<String> fieldNames = rel.getRowType().getFieldNames();
+      final RelOptTable relOptTable = rel.getTable();
+
+      double ndv = 1D;
+      for (Integer index : groupKey) {
+        final ColStatistics statistics = relOptTable.getColumnStatistics(fieldNames.get(index));
+        if (statistics == null) {
+          return null;
+        }
+
+        final Long cnt = statistics.getCountDistinct();
+        if (cnt == null) {
+          return null;
+        } else {
+          ndv *= cnt;
+        }
+      }
+      final Double rowCount = mq.getRowCount(rel);
+      return (rowCount == null ? ndv : Math.min(ndv, mq.getRowCount(rel))) * selectivity;
```

Comment:
The product of individual column NDVs seems an overestimate of the multiple column NDV. We should adjust it with the `RelMdUtil.numDistinctVals` method?